### PR TITLE
fix(cli): display internal commands in a subcommand #343

### LINKED ISSUE #343: Add an option to list internal commands
With the change to `internal__`, the plumbing commands are completely unaccessible from the `help` output. Which is sad, because then users cannot use them to experiment with unstable features.

Let’s add an option to list the internal commands and make it clear from the toplevel helptext that some commands are hidden by default.

### PR DESCRIPTION
**Pull request by bnjjj**
## Overview

Put each internal commands in a subcommands visible for all users.
close #343 

BTW should I remove the ping command ? (related to the TODO comment in file)

## Checklist

- [X] Updated the documentation (code documentation, command help, ...)
- [x] Tested the change (unit or integration tests)
- [X] Amended the changelog in `release.nix` (see `release.nix` for instructions)

### CODE REVIEW

File: src/cli.rs [hunk 1]
```diff
@@ -133,6 +120,22 @@ pub struct WatchOptions {
...
 }
 
+/// Sub-commands which lorri can execute for internal features
+#[derive(StructOpt, Debug)]
+pub enum Internal_ {
+    /// (internal) Used internally by `lorri shell`
+    #[structopt(name = "start_user_shell")]
+    StartUserShell_(StartUserShellOptions_),
+
+    /// (internal) Tell the lorri daemon to care about the current directory's project
```
Comment (Profpatsch) [inline src/cli.rs]:
```suggestion
    /// (plumbing) Tell the lorri daemon to care about the current directory's project
```

File: src/main.rs [hunk 2]
```diff
@@ -112,21 +109,28 @@ fn run_command(log: slog::Logger, opts: Arguments) -> OpResult {
...
-        }
-
         Command::Init => {
             let _guard = without_project();
             init::main(TRIVIAL_SHELL_SRC, DEFAULT_ENVRC)
         }
+
+        Command::Internal { command } => {
+            match command {
+                // TODO: remove
```
Comment (Profpatsch) [inline src/main.rs]: we can remove that TODO, I think ping is helpful for scripting.

**Comment by Profpatsch** [comment]
cc @curiousleo

**Comment by bnjjj** [comment]
@curiousleo My bad ! Thanks ! It's up to date now :)

File: MAINTAINERS.md [hunk 3]
```diff
@@ -8,8 +8,7 @@ non-negative integers without leading zeroes.
 ### Terminology
 
 lorri is a command line tool with multiple **subcommands**. Those subcommands
-are either _external_ or _internal_.  Internal subcommands' names start with
-"internal\_\_". There are _mandatory_ and _optional_ **command line options**,
+are either _external_ or _internal_. There are _mandatory_ and _optional_ **command line options**,
```
Comment (curiousleo) [inline MAINTAINERS.md]: I think it's worth mentioning how a user can tell: _internal_ subcommands are now all grouped in `lorri internal`.

File: src/ops/shell.rs [hunk 4]
```diff
@@ -56,7 +56,7 @@ pub fn main(project: Project, opts: ShellOptions) -> OpResult {
     bash_cmd
         .args(&[
             "-c",
-            "exec \"$1\" internal__start_user_shell --shell-path=\"$2\" --shell-file=\"$3\"",
+            "exec \"$1\" internal start_user_shell --shell-path=\"$2\" --shell-file=\"$3\"",
```
Comment (curiousleo) [inline src/ops/shell.rs]: Before this change in this file, the PR should have failed CI. So either
- there is a test that should have failed, but didn't - we need to investigate and fix (possibly in this PR)
- there is no test that covers this - we should add one (not in this PR)

Comment (bnjjj) [reply src/ops/shell.rs]: Ok I checked about tests. First it seems the execution of this command wasn't really checked. I added some check about the command's state just here https://github.com/target/lorri/pull/353/files#diff-9cbb4bd47e53163015bb66adcbc9db2cR24

Then I checked if a test existed and covered these lines of code. But it didn't, in fact the test never covered this function. So I took the decision to execute a test with `assert_cmd` to launch it as a real user in his prompt in order to cover these lines of code.

Let me know what do you think about that ?

File: src/ops/shell.rs [hunk 5]
```diff
@@ -68,7 +67,15 @@ pub fn main(project: Project, opts: ShellOptions) -> OpResult {
         ])
         .status()
         .expect("failed to execute bash");
-    Ok(())
+
+    if !status.success() {
+        Err(ExitError::panic(format!(
+            "cannot run lorri shell: fail to execute internal shell command (error: {})",
```
Comment (curiousleo) [inline src/ops/shell.rs]:
```suggestion
            "cannot run lorri shell: failed to execute internal shell command (error: {})",
```

File: tests/shell/main.rs [hunk 6]
```diff
@@ -1,18 +1,40 @@
+use assert_cmd::prelude::*;
```
Comment (curiousleo) [inline tests/shell/main.rs]: The rest of the codebase uses explicit `use` statements without wildcards. IMO it is worth having explicit `use` statements where possible to give people reading the code a better idea of what's in scope. Is it reasonable to turn this `use` statement into explicit ones?

Comment (bnjjj) [reply tests/shell/main.rs]: Ok updated :)

**Comment by bnjjj** [comment]
> Very nice, thanks for increasing the test coverage! Did you check that this test would have caught the wrong command in `shell.rs`?
> 
> https://github.com/target/lorri/blob/c38e0f09c36a1b27d86eb01ba5672fdffa950641/src/ops/shell.rs#L59

In fact the previous test didn't catch it. Now the new one catch it :)

File: tests/shell/main.rs [hunk 7]
```diff
@@ -1,18 +1,40 @@
...
+            "shell",
+            "--shell-file",
+            PathBuf::from(&project.nix_file)
+                .as_os_str()
+                .to_str()
+                .unwrap(),
+        ])
+        .current_dir(&tempdir)
+        .assert()
+        .success();
```
Comment (Profpatsch) [inline tests/shell/main.rs]: I’m not convinced that we need to include 8 new transitive dependencies for this one line. Let’s just do it manually.

Comment (bnjjj) [reply tests/shell/main.rs]: As it's in the dev-dependencies, is it really an issue ? It's only included for the test binary and not in the release binary. And I plan to add more tests about CLI so use this at different places.

Comment (Profpatsch) [reply tests/shell/main.rs]: The issue here is that every transitive dependency blows up CI time, and we run these tests on CI. Every dependency is a potential point of problems in the future, and has to carry its (transitive) weight.

> And I plan to add more tests about CLI so use this at different places.

Is it an undue amount of code to just use `Command` and pattern-match on the result like we do in the other tests?

Comment (bnjjj) [reply tests/shell/main.rs]: Ok 🙂 I updated the test

**Comment by bnjjj** [comment]
up

**Comment by Profpatsch** [comment]
Squashed & rebased on master. Merging once green.

**Pull request merged**